libraries/AP_ESC_Telem/AP_ESC_Telem_HobbyWing.cpp

```diff
@@ -0,0 +1,165 @@
+
+#if AP_HOBBYWING_PLATINUM_V4_ENABLED
+    // @Param: V4_MASK
+    // @DisplayName: HobbyWingESC v4 Channel Bitmask
+    // @Description: Mask of which channels are returning data on serial ports
+    // @Bitmask: 0:Channel1,1:Channel2,2:Channel3,3:Channel4,4:Channel5,5:Channel6,6:Channel7,7:Channel8,8:Channel9,9:Channel10,10:Channel11,11:Channel12,12:Channel13,13:Channel14,14:Channel15,15:Channel16, 16:Channel 17, 17: Channel 18, 18: Channel 19, 19: Channel 20, 20: Channel 21, 21: Channel 22, 22: Channel 23, 23: Channel 24, 24: Channel 25, 25: Channel 26, 26: Channel 27, 27: Channel 28, 28: Channel 29, 29: Channel 30, 30: Channel 31, 31: Channel 32
+    // @User: Advanced
+    // @RebootRequired: True
+    AP_GROUPINFO("V4_MASK",  12, AP_ESC_Telem_HobbyWing, channel_mask_v4, 0),
+
+    // @Param: V4_POLES
+    // @DisplayName: HobbyWingESC v4 Motor Poles
+    // @Description: This allows calculation of true RPM from ESC's eRPM. The default is 14.
+    // @Range: 1 127
+    // @User: Advanced
+    // @RebootRequired: True
+    AP_GROUPINFO("V4_POLES",  13, AP_ESC_Telem_HobbyWing, motor_poles_v4, 14),
+#endif  // AP_HOBBYWING_PLATINUM_V4_ENABLED
+
+#if AP_HOBBYWING_XROTOR_V4_ENABLED
+    // @Param: X4_MASK
+    // @DisplayName: HobbyWingESC XRotor v4 Channel Bitmask
+    // @Description: Mask of which channels are returning data on serial ports
+    // @Bitmask: 0:Channel1,1:Channel2,2:Channel3,3:Channel4,4:Channel5,5:Channel6,6:Channel7,7:Channel8,8:Channel9,9:Channel10,10:Channel11,11:Channel12,12:Channel13,13:Channel14,14:Channel15,15:Channel16, 16:Channel 17, 17: Channel 18, 18: Channel 19, 19: Channel 20, 20: Channel 21, 21: Channel 22, 22: Channel 23, 23: Channel 24, 24: Channel 25, 25: Channel 26, 26: Channel 27, 27: Channel 28, 28: Channel 29, 29: Channel 30, 30: Channel 31, 31: Channel 32
+    // @User: Advanced
+    // @RebootRequired: True
+    AP_GROUPINFO("X4_MASK",  22, AP_ESC_Telem_HobbyWing, channel_mask_x4, 0),
+
+    // @Param: X4_POLES
+    // @DisplayName: HobbyWingESC XRotor v4 Motor Poles
+    // @Description: This allows calculation of true RPM from ESC's eRPM. The default is 14.
+    // @Range: 1 127
+    // @User: Advanced
+    // @RebootRequired: True
+    AP_GROUPINFO("X4_POLES",  23, AP_ESC_Telem_HobbyWing, motor_poles_x4, 14),
+#endif  // AP_HOBBYWING_XROTOR_V4_ENABLED
+
+    AP_GROUPEND
+};
+
+// constructor
+AP_ESC_Telem_HobbyWing::AP_ESC_Telem_HobbyWing()
+{
+    AP_Param::setup_object_defaults(this, var_info);
+}
+
+void AP_ESC_Telem_HobbyWing::thread_main()
+{
+    // initialise all escs; this will set the UART up in the thread
+    for (uint8_t i=0; i<num_escs; i++) {
+        escs[i]->init();
+    }
+
+    while (true) {
+        hal.scheduler->delay_microseconds(1500);
+        for (uint8_t i=0; i<num_escs; i++) {
+            escs[i]->update();
+        }
+    }
+}
+
+void AP_ESC_Telem_HobbyWing::init()
+{
+    for (uint8_t i=0; i<32 && num_escs<ARRAY_SIZE(escs); i++) {
```

Comment:
Shouldn't this be dependent on whether you have support for 32 motors?

Reply:
Probably :-)  But the parameter is always going to be 32-bits regardless of your support for 32-bit, sooo...